### DISCUSSION TITLE
Detect subscription successfully started

### DISCUSSION THREAD

**Discussion by Amerehei** [enisdenjo/graphql-ws#528] [Q&A] [answered]
When we use [Observable](https://the-guild.dev/graphql/ws/recipes#client-usage-with-observable), using `error` we can detect that subscription is failed to star, But there is no way to detect that it's successfully started

How can I detect that the server response of subscription is received?

**Comment by enisdenjo** [comment]
Since there is no confirmation message on the protocol level, you can provide an initial subscription event that is sent as soon as the client subscribes.

**Reply by Amerehei** [reply]
What does `as soon as the client subscribes.` mean? I cannot detect it.
If I send a request to server and wait for a `next` event there are 3 scenarios

1. I get an error (We can handle it)
2. I get the `next` event (It's nice)
3. I lose the event, because the server get's the request sooner that I subscribe, So I will never get the `next` event

**Reply by Amerehei** [reply]
> there is no confirmation message on the protocol level

I think the protocol level should be changed, this is a very important requirement. How can I request for this?

**Reply by Amerehei** [reply]
> you can provide an initial subscription event that is sent as soon as the client subscribes.

I have no control over the server implementation

**Reply by enisdenjo** [reply]
>What does as soon as the client subscribes. mean?

You implement on GraphQL subscription level an event that is issued immediately.

> I lose the event, because the server get's the request sooner that I subscribe, So I will never get the next event

I don't understand this.

> I think the protocol level should be changed, this is a very important requirement. How can I request for this?

I don't think so, mainly because you can implement this on subscription level. If the WebSocket message is sent and the connection did not error out - you can be sure that the server has received the event.

Before we proceed with this discussion, what is your use case?

**Reply by Amerehei** [reply]
@enisdenjo What do you mean of "You implement on GraphQL subscription level an event that is issued immediately."
In the server side? I don't have access to server, I'm using a 3rd party server

> Before we proceed with this discussion, what is your use case?

In my frontend code, it is not safe to let user access the app before subscription, it can cause un-synced state.
I would like to show a loading UI and hide it after subscription

**Reply by enisdenjo** [reply]
> In my frontend code, it is not safe to let user access the app before subscription, it can cause un-synced state.
> I would like to show a loading UI and hide it after subscription

You can listen for the `SubscribeMessage` message on the client and hide the loading screen when it's sent. This should be safe since sending the `SubscribeMessage` message has two possible outcomes:

- Nothing happens - the subscribe request was successfully delivered to the server
- Error happens - you already handle it

```js
import { createClient, MessageType } from 'graphql-ws';

showLoadingUI();
const client = createClient({
  url: 'ws://localhost:4000/graphql',
  on: {
    message(message) {
      if (message.type === MessageType.Subscribe) {
        // optionally you can use the `message.payload` to hide the loading UI on specific query
        hideLoadingUI();
      }
    },
  },
});
```

**Reply by Amerehei** [reply]
@enisdenjo I know that your answer is correct. I faced a problem. I can get `connection_ack`. My problem is, when I send `subscribe` to the server there is either an error in case of error or no reply for the subscribe request in case of success.

Is it a server bug? You previously mentioned "here is no confirmation message on the protocol level" so how `message.type === MessageType.Subscribe` can work?

**Reply by enisdenjo** [reply]
> Is it a server bug?

Not a bug, `connection_ack` is for acknowledging the connection; meaning, the `connection_init` is valid and correct.

> You previously mentioned "here is no confirmation message on the protocol level" so how message.type === MessageType.Subscribe can work?

As I said [here](https://github.com/enisdenjo/graphql-ws/discussions/528#discussioncomment-7916682), if the WebSocket message is sent and the connection did not error out - you can be sure that the server has received the event.

So, if you send a subscribe message and the server does not reply with an error - you're good to go. Furthermore, even if the server is late to send the error message (because validation took too long or whatever) - you still show the error screen when you get the error, so you're good in that case too.

**Reply by Amerehei** [reply]
@enisdenjo This is not only a loading UI. I would like to resolve/reject a promise and I cannot resolve it immediately. If an error occurs in this case I cannot reject the promise. So there is no way to resolve the promise. 

> Is it a server bug? 
My question was not about the `connection_ack`. I asked I only get an acknowledging the connection. For the subscription there is no response so `message.type === MessageType.Subscribe` would never satisfy

I asked if we have `MessageType.Subscribe` and there is no confirmation message on the protocol level for subscription, How does it work?

**Reply by enisdenjo** [reply, marked as answer]
> I asked if we have MessageType.Subscribe and there is no confirmation message on the protocol level for subscription, How does it work?

There is no confirmation message for subscriptions, only for connections.

I just remembered, there is actually a way to acknowledge subscription - through pings. See the [recipe here](https://the-guild.dev/graphql/ws/recipes#ws-server-and-client-usage-with-subscription-acknowledgment). You will need to change the server though...

**Reply by Amerehei** [reply]
@enisdenjo Thanks a lot for sharing your knowledge
